### PR TITLE
[SPARK-11628][SQL] support column datatype of char(x) to recognize HiveChar

### PR DESCRIPTION
**Pull request by xguo27**
Can someone review my code to make sure I'm not missing anything? Thanks!

### CODE REVIEW

**Comment by cloud-fan** [comment]
Can't we use `ByteType` for char column?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 1]
```diff
@@ -96,6 +96,7 @@ private[sql] object JDBCRDD extends Logging {
       case java.sql.Types.TINYINT       => IntegerType
       case java.sql.Types.VARBINARY     => BinaryType
       case java.sql.Types.VARCHAR       => StringType
+      case java.sql.Types.CHAR          => StringType
```
Comment (Zhangshunyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: case java.sql.Types.CHAR          => StringType
Repeat code between line99 and line 65, plz delete this code.

**Comment by xguo27** [comment]
Hi Wenchen:

Can you elaborate on using ByteType for char a little more?

Ultimately, the difference between char(x) and varchar(x) is the fixed/variable length, which results in padding. So it's a good idea to keep the underlying type the same, right?

**Comment by cloud-fan** [comment]
Ah, sorry for the misunderstanding, I though you mean the one-byte `Char` in programing language...

Can you update your PR title to say that you are fixing hive part to recognize `HiveChar`? And please also add more tests to make sure we can read and write hive table with char(x) type columns.

**Comment by xguo27** [comment]
@cloud-fan I have added a few tests per your suggestion. Do they look good to you?

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by yhuai** [comment]
Thanks! Merging to master and branch 1.6.

**Comment by y304170319** [comment]
// hive varchar is treated as catalyst string, but hive varchar can't be pushed down.

**Comment by y304170319** [comment]
"// hive varchar is treated as catalyst string, but hive varchar can't be pushed down."
In HiveShim, there is comment above, why varchar can't be pushed down?
Do you know ?  Thanks.

**Pull request closed without merging**